File: src/lib/mentions.ts
/** Shared @imgN reference-tag logic (used by client UI and server routes). */

export interface LabeledRef {
  tag: string; // "@img1"
  index: number; // 1-based
  dataUrl: string; // data:image/...;base64,...
}

/** Match @img1, @IMG2, etc. (ad-hoc one-off uploads). */
export const MENTION_REGEX = /@img(\d+)/gi;

/** Match @vid1, @VID2 … — attached reference CLIPS (video-to-video).
 *  A separate namespace from @imgN because the two travel to the provider by
 *  completely different routes (inline base64 vs presigned URL). */
export const VIDEO_MENTION_REGEX = /@vid(\d+)/gi;

/** Match any @tag token: ad-hoc @imgN OR a named asset slug like @priya. */
export const TAG_REGEX = /@([a-z][a-z0-9_-]*)/gi;

/** True for ad-hoc upload tags (@img1, @img2 …) vs named asset slugs. */
export function isImgTag(slug: string): boolean {
  return /^img\d+$/i.test(slug);
}

/** True for attached-clip tags (@vid1, @vid2 …). */
export function isVidTag(slug: string): boolean {
  return /^vid\d+$/i.test(slug);
}

/**
 * Named asset slugs referenced in a prompt (e.g. @priya, @red-lehenga), in
 * first-appearance order, excluding the ad-hoc @imgN tokens.
 */
export function parseAssetSlugs(prompt: string): string[] {
  const seen = new Set<string>();
  const order: string[] = [];
  const re = new RegExp(TAG_REGEX);
  let m: RegExpExecArray | null;
  while ((m = re.exec(prompt))) {
    const slug = m[1].toLowerCase();
    // @vidN is a clip tag, not an asset slug. Without this it was looked up as
    // a saved asset named "vid1", found nothing, and silently stayed in the
    // prompt as ordinary text — which is why typing @vid1 appeared to do
    // nothing at all.
    if (isImgTag(slug) || isVidTag(slug) || seen.has(slug)) continue;
    seen.add(slug);
    order.push(slug);
  }
  return order;
}

/** Unique 1-based indices referenced by @imgN tokens, in ascending order. */
export function parseMentionIndices(prompt: string): number[] {
  const set = new Set<number>();
  const re = new RegExp(MENTION_REGEX);
  let m: RegExpExecArray | null;
  while ((m = re.exec(prompt))) {
    const n = parseInt(m[1], 10);
    if (n >= 1) set.add(n);
  }
  return [...set].sort((a, b) => a - b);
}

/**
 * Decide which uploaded images to actually send to the model.
 * - If the prompt tags images (@img1 …), send only those (the tag = intent).
 * - If the prompt tags none, fall back to sending all uploads.
 * Out-of-range tags (e.g. @img9 with 2 uploads) are ignored.
 */
/** Unique 1-based indices referenced by @vidN tokens, ascending. */
export function parseVideoMentionIndices(prompt: string): number[] {
  const set = new Set<number>();
  const re = new RegExp(VIDEO_MENTION_REGEX);
  let m: RegExpExecArray | null;
  while ((m = re.exec(prompt))) {
    const n = parseInt(m[1], 10);
    if (n >= 1) set.add(n);
  }
  return [...set].sort((a, b) => a - b);
}

/**
 * Which attached clips to actually send, mirroring resolveReferences: an
 * explicit @vidN tag is intent, no tags means send them all.
 */
export function resolveVideoReferences(
  prompt: string,
  clips: string[]
): string[] {
  if (!clips?.length) return [];
  const tagged = parseVideoMentionIndices(prompt).filter((n) => n <= clips.length);
  const indices = tagged.length ? tagged : clips.map((_, i) => i + 1);
  return indices.map((n) => clips[n - 1]);
}

export function resolveReferences(
  prompt: string,
  uploads: string[]
): LabeledRef[] {
  if (!uploads?.length) return [];
  const tagged = parseMentionIndices(prompt).filter((n) => n <= uploads.length);
  const indices = tagged.length ? tagged : uploads.map((_, i) => i + 1);
  return indices.map((n) => ({
    tag: `@img${n}`,
    index: n,
    dataUrl: uploads[n - 1],
  }));
}

File: backend/apps/generation/mentions.py
"""Direct port of src/lib/mentions.js — shared @imgN reference-tag logic."""

import re

MENTION_RE = re.compile(r"@img(\d+)", re.IGNORECASE)
VIDEO_MENTION_RE = re.compile(r"@vid(\d+)", re.IGNORECASE)
TAG_RE = re.compile(r"@([a-z][a-z0-9_-]*)", re.IGNORECASE)


def is_img_tag(slug: str) -> bool:
    return bool(re.match(r"^img\d+$", slug, re.IGNORECASE))


def is_vid_tag(slug: str) -> bool:
    return bool(re.match(r"^vid\d+$", slug, re.IGNORECASE))


def parse_asset_slugs(prompt: str) -> list[str]:
    """Named asset slugs referenced in a prompt, first-appearance order,
    excluding ad-hoc @imgN tokens. @vidN is a clip tag, not an asset slug —
    without that exclusion it would be looked up as a saved asset named
    "vid1", found nothing, and silently stayed in the prompt as text."""
    seen: set[str] = set()
    order: list[str] = []
    for m in TAG_RE.finditer(prompt):
        slug = m.group(1).lower()
        if is_img_tag(slug) or is_vid_tag(slug) or slug in seen:
            continue
        seen.add(slug)
        order.append(slug)
    return order


def parse_mention_indices(prompt: str) -> list[int]:
    indices = {int(m.group(1)) for m in MENTION_RE.finditer(prompt) if int(m.group(1)) >= 1}
    return sorted(indices)


def parse_video_mention_indices(prompt: str) -> list[int]:
    indices = {int(m.group(1)) for m in VIDEO_MENTION_RE.finditer(prompt) if int(m.group(1)) >= 1}
    return sorted(indices)


def resolve_video_references(prompt: str, clips: list[str]) -> list[str]:
    """Which attached clips to actually send: an explicit @vidN tag is
    intent, no tags means send them all."""
    if not clips:
        return []
    tagged = [n for n in parse_video_mention_indices(prompt) if n <= len(clips)]
    indices = tagged if tagged else list(range(1, len(clips) + 1))
    return [clips[n - 1] for n in indices]


def resolve_references(prompt: str, uploads: list[str]) -> list[dict]:
    """Which uploaded images to actually send: an explicit @imgN tag is
    intent (send only those); no tags means send all uploads. Out-of-range
    tags (e.g. @img9 with 2 uploads) are ignored."""
    if not uploads:
        return []
    tagged = [n for n in parse_mention_indices(prompt) if n <= len(uploads)]
    indices = tagged if tagged else list(range(1, len(uploads) + 1))
    return [{"tag": f"@img{n}", "index": n, "dataUrl": uploads[n - 1]} for n in indices]


def renumber_img_mentions(prompt: str, mapping: list[int | None]) -> str:
    """Rewrites @imgN tokens so they keep pointing at the same physical
    image after the upload array is reordered. mapping[oldIndex] (0-based)
    is the image's new 0-based index; an old index missing from mapping
    (out of range) is left untouched."""

    def replace(m: re.Match) -> str:
        old_index = int(m.group(1)) - 1
        new_index = mapping[old_index] if 0 <= old_index < len(mapping) else None
        return m.group(0) if new_index is None else f"@img{new_index + 1}"

    return MENTION_RE.sub(replace, prompt)
